storage/grpc_client.go

```diff
@@ -2383,77 +2379,88 @@ func checkCanceled(err error) error {
 	return err
 }
 
-func (w *gRPCWriter) initializeRetryConfig() {
-	if w.attempts == 0 {
-		w.attempts = 1
+type uploadBufferRetryConfig struct {
+	attempts     int
+	invocationID string
+	config       *retryConfig
+	lastErr      error
+}
 
-		if w.settings.retry == nil {
-			w.settings.retry = defaultRetry
-		}
+func newUploadBufferRetryConfig(settings *settings) *uploadBufferRetryConfig {
+	config := settings.retry
 
-		if w.settings.retry.shouldRetry == nil {
-			w.settings.retry.shouldRetry = ShouldRetry
-		}
+	if config == nil {
+		config = defaultRetry.clone()
+	}
 
-		w.invocationID = uuid.New().String()
+	if config.shouldRetry == nil {
+		config.shouldRetry = ShouldRetry
+	}
 
-		if w.settings.retry.backoff != nil {
-			w.backoff.Multiplier = w.settings.retry.backoff.Multiplier
-			w.backoff.Initial = w.settings.retry.backoff.Initial
-			w.backoff.Max = w.settings.retry.backoff.Max
-		}
+	if config.backoff == nil {
+		config.backoff = &gaxBackoff{}
+	} else {
+		config.backoff.SetMultiplier(settings.retry.backoff.GetMultiplier())
+		config.backoff.SetInitial(settings.retry.backoff.GetInitial())
+		config.backoff.SetMax(settings.retry.backoff.GetMax())
+	}
+
+	return &uploadBufferRetryConfig{
+		attempts:     1,
+		invocationID: uuid.New().String(),
+		config:       config,
 	}
 }
 
-// shouldRetry determines if a retry is necessary and if so waits the appropriate
-// amount of time. It returns true if the error is retryable or the error to be
-// surfaced to the user if not.
-func (w *gRPCWriter) shouldRetry(ctx context.Context, err error) (bool, error) {
+// retriable determines if a retry is necessary and if so returns a nil error;
+// otherwise it returns the error to be surfaced to the user.
+func (retry *uploadBufferRetryConfig) retriable(ctx context.Context, err error) error {
```

Comment:
It seems a little odd to have this return the error to be resurfaced. Maybe just have it return a bool, and you can use retry.lastErr to set the error to return if need be?

Also not a big deal if you want to leave this as-is for now, since we'll be overwriting with the refactored version anyway.

Reply:
That's fair; I think I may have been too deep in this. Although, this returns a formatted error for when max attempts are reached, which you wouldn't know outside this method unless you returned two bools... but I guess adding the number of attempts to the error may be useful for all cases anyway. I'll leave as-is and consider this feedback if relevant to the other version.